README.md
# admob_app

A AdMob Application Example

## Getting Started

For help getting started with Flutter, view our online
[documentation](https://flutter.io/).


CHANGELOG.md
## 0.1.0 
  June 24, ‎2019. 
- Singleton Design Pattern


lib/ads_test.dart
import 'package:flutter/material.dart';

import 'package:firebase_admob/firebase_admob.dart';

import 'package:flutter_test/flutter_test.dart';

import 'package:ads/ads.dart';

//import '../example/main.dart';
import 'main.dart';

Key _showBanner = ValueKey<String>('SHOW BANNER');
Key _removeBanner = ValueKey<String>('REMOVE BANNER');
Key _showFullScreen = ValueKey<String>('SHOW INTERSTITIAL');
Key _showVideo = ValueKey<String>('SHOW REWARDED VIDEO');

/// Names the last event triggered.
String _event = '';

void main() {
  testWidgets('Test Rewarded Video', (WidgetTester tester) async {
    Key key = UniqueKey();

    await tester.pumpWidget(MaterialApp(
      home: MyApp(
        initOption: 1,
        key: key,
      ),
    ));

    Ads.video.startedListener = () {
      _event = 'started';
    };

    Ads.video.completedListener = () {};

    // Play a video ad.
    await tester.tap(find.byKey(_showVideo));
    await tester.pumpAndSettle();
//    await tester.idle();

    Text _coins = tester.widget(find.byKey(ValueKey<String>('COINS')));
    String value = _coins.data;

    // Verify the current counter.
    expect(find.text(value), findsOneWidget);

    expect(Ads.keywords.contains('the'), isTrue);
    expect(Ads.contentUrl, isNull);
    expect(Ads.childDirected, isFalse);
    expect(Ads.testDevices, isEmpty);
  });

  String where = '';

//  testWidgets('Header adds todo', (WidgetTester tester) async {
//    await tester.pumpWidget(MaterialApp(
//      home: Material(
//        child: MyApp(
//          initOption: 2,
//        ),
//      ),
//    ));
//
//    Ads.banner.loadedListener = () {
//      where = 'loaded';
//    };
//
//    Ads.banner.clickedListener = () {
//      where = 'clicked';
//    };
//
//    Ads.banner.openedListener = () {
//      where = 'opened';
//    };
//
//    Ads.banner.leftAppListener = () {
//      where = 'left';
//    };
//
//    Ads.banner.closedListener = () {
//      where = 'closed';
//    };
//
//    Ads.video.openedListener = () {
//      where = 'opened';
//    };
//
//    await tester.tap(find.byKey(_showBanner));
//    await tester.pumpAndSettle();
//
////    expect(where, equals('MobileAdEvent.opened'));
//
//
////    Ads.clearVideoListeners();
//
////    Ads.videoListener(VideoEventListener listener) =>
//
//    expect(Ads.appId, equals(FirebaseAdMob.testAppId));
//    expect(Ads.keywords.contains('ibm'), isTrue);
//    expect(Ads.keywords.contains('computers'), isTrue);
//    expect(Ads.contentUrl, equals('http://www.ibm.com'));
//    expect(Ads.childDirected, isFalse);
//    expect(Ads.testDevices.contains('Samsung_Galaxy_SII_API_26:5554'), isTrue);
//    expect(Ads.testing, isFalse);
////    Ads.banner.clearAll();
//
//  });
//
//  testWidgets('Header adds todo', (WidgetTester tester) async {
//    Key key = UniqueKey();
//
//    await tester.pumpWidget(MaterialApp(
//      home: Material(
//        child: MyApp(
//          initOption: 3,
//          key: key,
//        ),
//      ),
//    ));
//
//    await tester.tap(find.byType(BannerAd));
//
//    expect(_event, equals('MobileAdEvent.leftApplication'));
//
//    expect(Ads.appId, equals(FirebaseAdMob.testAppId));
//    expect(Ads.keywords.contains('cats'), isTrue);
//    expect(Ads.keywords.contains('dogs'), isTrue);
//    expect(Ads.contentUrl, equals('http://www.animalsaspets.com'));
//    expect(Ads.childDirected, isFalse);
//    expect(Ads.testDevices.contains('Samsung_Galaxy_SII_API_26:5554'), isTrue);
//    expect(Ads.testing, isTrue);
//  });
//
//  testWidgets('Header adds todo', (WidgetTester tester) async {
//    Key key = UniqueKey();
//
//    await tester.pumpWidget(MaterialApp(
//      home: Material(
//        child: MyApp(
//          initOption: 4,
//          key: key,
//        ),
//      ),
//    ));
//
//    expect(Ads.bannerAd.adUnitId, equals(FirebaseAdMob.testAppId));
//    expect(Ads.bannerAd.targetingInfo.keywords.contains('android'), isTrue);
//    expect(Ads.bannerAd.targetingInfo.keywords.contains('flutter'), isTrue);
//    expect(Ads.bannerAd.targetingInfo.contentUrl,
//        equals('http://www.andrioussolutions.com'));
//    expect(Ads.bannerAd.targetingInfo.childDirected, isFalse);
//    expect(
//        Ads.bannerAd.targetingInfo.testDevices
//            .contains('Samsung_Galaxy_SII_API_26:5554'),
//        isTrue);
//
//    expect(Ads.fullScreenAd.adUnitId, equals(FirebaseAdMob.testAppId));
//    expect(Ads.fullScreenAd.targetingInfo.keywords.contains('dart'), isTrue);
//    expect(Ads.fullScreenAd.targetingInfo.keywords.contains('flutter'), isTrue);
//    expect(Ads.fullScreenAd.targetingInfo.contentUrl,
//        equals('http://www.fluttertogo.com'));
//    expect(Ads.fullScreenAd.targetingInfo.childDirected, isFalse);
//    expect(
//        Ads.fullScreenAd.targetingInfo.testDevices
//            .contains('Samsung_Galaxy_SII_API_26:5554'),
//        isTrue);
//
//    expect(Ads.videoAd.adUnitId, equals(FirebaseAdMob.testAppId));
//    expect(Ads.videoAd.targetingInfo.keywords.contains('dart'), isTrue);
//    expect(Ads.videoAd.targetingInfo.keywords.contains('java'), isTrue);
//    expect(
//        Ads.videoAd.targetingInfo.contentUrl, equals('http://www.publang.org'));
//    expect(Ads.videoAd.targetingInfo.childDirected, isFalse);
//    expect(
//        Ads.videoAd.targetingInfo.testDevices
//            .contains('Samsung_Galaxy_SII_API_26:5554'),
//        isFalse);
//    expect(Ads.videoAd.listener, isNotNull);
//  });
//
//  testWidgets('Header adds todo', (WidgetTester tester) async {
//    Key key = UniqueKey();
//
//    await tester.pumpWidget(MaterialApp(
//      home: Material(
//        child: MyApp(
//          initOption: 5,
//          key: key,
//        ),
//      ),
//    ));
//
//    expect(Ads.keywords.contains('cats'), isTrue);
//    expect(Ads.keywords.contains('dogs'), isTrue);
//    expect(Ads.contentUrl, equals('http://www.animalsaspets.com'));
//    expect(Ads.childDirected, isFalse);
//    expect(Ads.testDevices.contains('Samsung_Galaxy_SII_API_26:5554'), isTrue);
//    expect(Ads.testing, isTrue);
//  });
//
//  testWidgets('Header adds todo', (WidgetTester tester) async {
//    Key key = UniqueKey();
//
//    await tester.pumpWidget(MaterialApp(
//      home: Material(
//        child: MyApp(
//          initOption: 6,
//          key: key,
//        ),
//      ),
//    ));
//  });
//
//  testWidgets('Header adds todo', (WidgetTester tester) async {
//    Key key = UniqueKey();
//
//    await tester.pumpWidget(MaterialApp(
//      home: Material(
//        child: MyApp(
//          initOption: 7,
//          key: key,
//        ),
//      ),
//    ));
//  });
//
//  testWidgets('Header adds todo', (WidgetTester tester) async {
//    Key key = UniqueKey();
//
//    await tester.pumpWidget(MaterialApp(
//      home: Material(
//        child: MyApp(
//          initOption: 8,
//          key: key,
//        ),
//      ),
//    ));
//  });
//
//  testWidgets('Header adds todo', (WidgetTester tester) async {
//    Key key = UniqueKey();
//
//    await tester.pumpWidget(MaterialApp(
//      home: Material(
//        child: MyApp(
//          initOption: 9,
//          key: key,
//        ),
//      ),
//    ));
//  });
//
//  testWidgets('Header adds todo', (WidgetTester tester) async {
//    Key key = UniqueKey();
//
//    await tester.pumpWidget(MaterialApp(
//      home: Material(
//        child: MyApp(
//          initOption: 10,
//          key: key,
//        ),
//      ),
//    ));
//  });
}


test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

//import '../../../libs/ads/example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
//    // Build our app and trigger a frame.
//    await tester.pumpWidget(new MyApp());
//
//    // Verify that our counter starts at 0.
//    expect(find.text('0'), findsOneWidget);
//    expect(find.text('1'), findsNothing);
//
//    // Tap the '+' icon and trigger a frame.
//    await tester.tap(find.byIcon(Icons.add));
//    await tester.pump();
//
//    // Verify that our counter has incremented.
//    expect(find.text('0'), findsNothing);
//    expect(find.text('1'), findsOneWidget);
  });
}


lib/example01.dart
import 'package:flutter/material.dart';
import 'dart:io' show Platform;
import 'package:firebase_admob/firebase_admob.dart';

import 'package:ads/ads.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  
  final String appId = Platform.isAndroid
      ? 'ca-app-pub-3940256099942544~3347511713'
      : 'ca-app-pub-3940256099942544~1458002511';

  final String bannerUnitId = Platform.isAndroid
      ? 'ca-app-pub-3940256099942544/6300978111'
      : 'ca-app-pub-3940256099942544/2934735716';

  @override
  void initState() {
    super.initState();

    /// Assign a listener.
    var eventListener = (MobileAdEvent event) {
      if (event == MobileAdEvent.clicked) {
        print("The opened ad is clicked on.");
      }
    };

    Ads.init(
      appId,
      bannerUnitId: bannerUnitId,
      keywords: <String>['ibm', 'computers'],
      contentUrl: 'http://www.ibm.com',
      childDirected: false,
      testDevices: ['Samsung_Galaxy_SII_API_26:5554'],
      testing: false,
      listener: eventListener,
    );

    Ads.showBannerAd();
  }

  @override
  void dispose() {
    Ads.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('AdMob Banner Ad Example'),
        ),
        body: SingleChildScrollView(
          child: Center(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              mainAxisSize: MainAxisSize.min,
              children: <Widget>[
                RaisedButton(
                    key: ValueKey<String>('SHOW BANNER'),
                    child: const Text('SHOW BANNER'),
                    onPressed: () {
                      Ads.showBannerAd(state: this, anchorOffset: null);
                    }),
                RaisedButton(
                    key: ValueKey<String>('REMOVE BANNER'),
                    child: const Text('REMOVE BANNER'),
                    onPressed: () {
                      Ads.hideBannerAd();
                    }),
              ].map((Widget button) {
                return Padding(
                  padding: const EdgeInsets.symmetric(vertical: 16.0),
                  child: button,
                );
              }).toList(),
            ),
          ),
        ),
      ),
    );
  }
}


lib/example02.dart
import 'package:flutter/material.dart';
import 'dart:io' show Platform;
import 'package:firebase_admob/firebase_admob.dart';

import 'package:ads/ads.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  int _coins = 0;

  final String appId = Platform.isAndroid
      ? 'ca-app-pub-3940256099942544~3347511713'
      : 'ca-app-pub-3940256099942544~1458002511';

  final String bannerUnitId = Platform.isAndroid
      ? 'ca-app-pub-3940256099942544/6300978111'
      : 'ca-app-pub-3940256099942544/2934735716';

  final String screenUnitId = Platform.isAndroid
      ? 'ca-app-pub-3940256099942544/1033173712'
      : 'ca-app-pub-3940256099942544/4411468910';

  final String videoUnitId = Platform.isAndroid
      ? 'ca-app-pub-3940256099942544/5224354917'
      : 'ca-app-pub-3940256099942544/1712485313';

  @override
  void initState() {
    super.initState();

    /// Assign a listener.
    var eventListener = (MobileAdEvent event) {
      if (event == MobileAdEvent.opened) {
        print("The opened ad is clicked on.");
      }
    };

    Ads.init(
      appId,
      bannerUnitId: bannerUnitId,
      screenUnitId: screenUnitId,
      keywords: <String>['ibm', 'computers'],
      contentUrl: 'http://www.ibm.com',
      childDirected: false,
      testDevices: ['Samsung_Galaxy_SII_API_26:5554'],
      testing: false,
      listener: eventListener,
    );

    Ads.setVideoAd(
      adUnitId: videoUnitId,
      keywords: ['dart', 'java'],
      contentUrl: 'http://www.publang.org',
      childDirected: true,
      testDevices: null,
      listener: (RewardedVideoAdEvent event,
          {String rewardType, int rewardAmount}) {
        print("The ad was sent a reward amount.");
        setState(() {
          _coins += rewardAmount;
        });
      },
    );

    /// 28 event listeners are initialized.
    _setEventListeners();

    Ads.showBannerAd();
  }

  @override
  void dispose() {
    Ads.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('AdMob Ad Examples'),
        ),
        body: SingleChildScrollView(
          child: Center(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              mainAxisSize: MainAxisSize.min,
              children: <Widget>[
                RaisedButton(
                    key: ValueKey<String>('SHOW BANNER'),
                    child: const Text('SHOW BANNER'),
                    onPressed: () {
                      Ads.showBannerAd(state: this, anchorOffset: null);
                    }),
                RaisedButton(
                    key: ValueKey<String>('REMOVE BANNER'),
                    child: const Text('REMOVE BANNER'),
                    onPressed: () {
                      Ads.hideBannerAd();
                    }),
                RaisedButton(
                  key: ValueKey<String>('SHOW INTERSTITIAL'),
                  child: const Text('SHOW INTERSTITIAL'),
                  onPressed: () {
                    Ads.showFullScreenAd(state: this);
                  },
                ),
                RaisedButton(
                  key: ValueKey<String>('SHOW REWARDED VIDEO'),
                  child: const Text('SHOW REWARDED VIDEO'),
                  onPressed: () {
                    Ads.showVideoAd(state: this);
                  },
                ),
                Text(
                  "You have $_coins coins.",
                  key: ValueKey<String>('COINS'),
                ),
              ].map((Widget button) {
                return Padding(
                  padding: const EdgeInsets.symmetric(vertical: 16.0),
                  child: button,
                );
              }).toList(),
            ),
          ),
        ),
      ),
    );
  }
}


void _setEventListeners(){

  Ads.eventListener = (MobileAdEvent event) {
    switch (event) {
      case MobileAdEvent.loaded:
        print("An ad has loaded successfully in memory.");
        break;
      case MobileAdEvent.failedToLoad:
        print("The ad failed to load into memory.");
        break;
      case MobileAdEvent.clicked:
        print("The opened ad was clicked on.");
        break;
      case MobileAdEvent.impression:
        print("The user is still looking at the ad. A new ad came up.");
        break;
      case MobileAdEvent.opened:
        print("The Ad is now open.");
        break;
      case MobileAdEvent.leftApplication:
        print("You've left the app after clicking the Ad.");
        break;
      case MobileAdEvent.closed:
        print("You've closed the Ad and returned to the app.");
        break;
      default:
        print("There's a 'new' MobileAdEvent?!");
    }
  };

  Ads.bannerListener = (MobileAdEvent event) {
    switch (event) {
      case MobileAdEvent.loaded:
        print("An ad has loaded successfully in memory.");
        break;
      case MobileAdEvent.failedToLoad:
        print("The ad failed to load into memory.");
        break;
      case MobileAdEvent.clicked:
        print("The opened ad was clicked on.");
        break;
      case MobileAdEvent.impression:
        print("The user is still looking at the ad. A new ad came up.");
        break;
      case MobileAdEvent.opened:
        print("The Ad is now open.");
        break;
      case MobileAdEvent.leftApplication:
        print("You've left the app after clicking the Ad.");
        break;
      case MobileAdEvent.closed:
        print("You've closed the Ad and returned to the app.");
        break;
      default:
        print("There's a 'new' MobileAdEvent?!");
    }
  };

  Ads.screenListener = (MobileAdEvent event) {
    switch (event) {
      case MobileAdEvent.loaded:
        print("An ad has loaded successfully in memory.");
        break;
      case MobileAdEvent.failedToLoad:
        print("The ad failed to load into memory.");
        break;
      case MobileAdEvent.clicked:
        print("The opened ad was clicked on.");
        break;
      case MobileAdEvent.impression:
        print("The user is still looking at the ad. A new ad came up.");
        break;
      case MobileAdEvent.opened:
        print("The Ad is now open.");
        break;
      case MobileAdEvent.leftApplication:
        print("You've left the app after clicking the Ad.");
        break;
      case MobileAdEvent.closed:
        print("You've closed the Ad and returned to the app.");
        break;
      default:
        print("There's a 'new' MobileAdEvent?!");
    }
  };

  Ads.videoListener =
      (RewardedVideoAdEvent event, {String rewardType, int rewardAmount}) {
    switch (event) {
      case RewardedVideoAdEvent.loaded:
        print("An ad has loaded successfully in memory.");
        break;
      case RewardedVideoAdEvent.failedToLoad:
        print("The ad failed to load into memory.");
        break;
      case RewardedVideoAdEvent.opened:
        print("The ad is now open.");
        break;
      case RewardedVideoAdEvent.leftApplication:
        print("You've left the app after clicking the Ad.");
        break;
      case RewardedVideoAdEvent.closed:
        print("You've closed the Ad and returned to the app.");
        break;
      case RewardedVideoAdEvent.rewarded:
        print("The ad has sent a reward amount.");
        break;
      case RewardedVideoAdEvent.started:
        print("You've just started playing the Video ad.");
        break;
      case RewardedVideoAdEvent.completed:
        print("You've just finished playing the Video ad.");
        break;
      default:
        print("There's a 'new' RewardedVideoAdEvent?!");
    }
  };

  Ads.banner.loadedListener = () {
    print("An ad has loaded successfully in memory.");
  };

  Ads.banner.failedListener = () {
    print("An ad failed to load into memory.");
  };

  Ads.banner.clickedListener = () {
    print("The opened ad is clicked on.");
  };

  Ads.banner.impressionListener = () {
    print("The user is still looking at the ad. A new ad came up.");
  };

  Ads.banner.openedListener = () {
    print("You've closed an ad and returned to your app.");
  };

  Ads.banner.leftAppListener = () {
    print("You left the app and gone to the ad's website.");
  };

  Ads.banner.impressionListener = () {
    print("The user is still looking at the ad. A new ad came up.");
  };

  Ads.banner.closedListener = () {
    print("You've closed an ad and returned to your app.");
  };

  Ads.screen.loadedListener = () {
    print("An ad has loaded into memory.");
  };

  Ads.screen.failedListener = () {
    print("An ad has failed to load in memory.");
  };

  Ads.screen.clickedListener = () {
    print("The opened ad was clicked on.");
  };

  Ads.screen.impressionListener = () {
    print("You've clicked on a link in the open ad.");
  };

  Ads.screen.openedListener = () {
    print("The ad has opened.");
  };

  Ads.screen.leftAppListener = () {
    print("The user has left the app and gone to the opened ad.");
  };

  Ads.screen.closedListener = () {
    print("The ad has been closed. The user returns to the app.");
  };

  Ads.video.loadedListener = () {
    print("An ad has loaded in memory.");
  };

  Ads.video.failedListener = () {
    print("An ad has failed to load in memory.");
  };

  Ads.video.clickedListener = () {
    print("An ad has been clicked on.");
  };

  Ads.video.openedListener = () {
    print("An ad has been opened.");
  };

  Ads.video.leftAppListener = () {
    print("You've left the app to view the video.");
  };

  Ads.video.closedListener = () {
    print("The video has been closed.");
  };

  Ads.video.rewardedListener = (String rewardType, int rewardAmount) {
    print("The ad was sent a reward amount.");
  };

  Ads.video.startedListener = () {
    print("You've just started playing the Video ad.");
  };

  Ads.video.completedListener = () {
    print("You've just finished playing the Video ad.");
  };

  List<String> two = Ads.keywords;
  String three = Ads.contentUrl;
  bool seven = Ads.childDirected;
  List<String> eight = Ads.testDevices;
  print(two);
}
